### PR TITLE
Wizard compatibility

### PR DESCRIPTION
**Pull request by Pringels**

### CODE REVIEW

**Comment by hedleyroos** [comment]
I don't understand how this would work. The `cached_result` value is set on an instance of ReCaptchaField but field instances are not persistent across requests. As you navigate from one page to another in a wizard that field is recreated. For it to work you need to store the cached_result somewhere else. You can't use the caching backend because (1) some systems round robin requests and if they don't have a shared cache backend it will get lost and (2) caches are inherently volatile.

I suggest storing the cache result on the user's session. You already use thread locals (`if 'request' in f.f_locals:`) to retrieve the request and can access the session from that. 

You will need to find a way that the cached result is valid only for this wizard. That will prove tricky because the recaptchafield has no idea of the flow happening in the wizard it lives inside. Easiest is probably to use request.path_info as part of the key you store in the session, eg. `rcf-/some/path/cached-result` and bake an expiry time into the value.

File: README.rst [hunk 1]
```diff
@@ -3,10 +3,10 @@ Django reCAPTCHA
 **Django reCAPTCHA form field/widget integration app.**
 
 .. image:: https://travis-ci.org/praekelt/django-recaptcha.svg?branch=develop
-    :target: https://travis-ci.org/praekelt/django-recaptcha
+:target: https://travis-ci.org/praekelt/django-recaptcha
```
Comment (adamchainz) [inline README.rst]: why?

Comment (Pringels) [inline README.rst]: My IDE's auto reformatter. Will revert

File: README.rst [hunk 2]
```diff
@@ -100,6 +100,28 @@ field, containing a dictionary of options. For example:
...
+
+This will cache the successful captcha validation response and prevent subsequent validation
+failures when the wizard reruns validation on your forms.
+
+The result will be cached for 10 minutes by default.
+If you require a custom expiration period, add a ``RECAPTCHA_WIZARD_EXPIRY`` setting, for example:
+
+.. code-block:: python
+
+   RECAPTCHA_WIZARD_EXPIRY = 30
```
Comment (hedleyroos) [inline README.rst]: Document that it's minutes, not seconds. The Django convention is to specify everything in seconds by the way.

File: captcha/fields.py [hunk 3]
```diff
@@ -23,7 +26,7 @@ class ReCaptchaField(forms.CharField):
     }
 
     def __init__(self, public_key=None, private_key=None, use_ssl=None,
-                 attrs=None, *args, **kwargs):
+                 wizard=None, attrs=None, *args, **kwargs):
```
Comment (hedleyroos) [inline captcha/fields.py]: I missed this before. If you introduce a new argument then any existing code that subclasses this class will break. You will have to do

    wizard = kwargs.get("wizard", None)

The nice thing about kwargs is you can change your method signatures and not break other code.

File: captcha/fields.py [hunk 4]
```diff
@@ -39,24 +42,54 @@ def __init__(self, public_key=None, private_key=None, use_ssl=None,
...
+                expires_time = request.session[key]
+                return expires_time > time.time()
+
+        return False
+
+    def set_cached_result(self):
+        request = self.get_request_object()
+        if request:
+            key = 'rcf-{path}cached-result'.format(path=request.path_info)
+            now = datetime.datetime.now()
```
Comment (hedleyroos) [inline captcha/fields.py]: I'm worried about times without timezones. You should make all your times tz aware, but explicitly choose UTC as your zone. Makes for easy debugging. However, just create an issue for that. At worst the captcha is cached a little longer or shorter than expected if things go wrong.

Comment (Pringels) [inline captcha/fields.py]: Issue created

Comment (Pringels) [inline captcha/fields.py]: Although I'm not entirely sure why timezones would be an issue in this case. Surely the time delta only ever gets calculated using times defined within the scope of a specific session?

Comment (hedleyroos) [inline captcha/fields.py]: You're probably right. My paranoia at being bitten by timezones is taking over :) I was worried that the code delegates to a Django internal that does consider the timezone as defined in settings but it doesn't.

File: captcha/fields.py [hunk 5]
```diff
@@ -39,24 +42,54 @@ def __init__(self, public_key=None, private_key=None, use_ssl=None,
...
+        if request:
+            remote_ip = request.META.get('REMOTE_ADDR', '')
+            forwarded_ip = request.META.get('HTTP_X_FORWARDED_FOR', '')
+            ip = remote_ip if not forwarded_ip else forwarded_ip
+            return ip
+
+    def get_cached_result(self):
+        request = self.get_request_object()
+        if request:
+            key = 'rcf-{path}cached-result'.format(path=request.path_info)
```
Comment (Pringels) [inline captcha/fields.py]: Probably a good idea. Will add it tomorrow.

File: captcha/fields.py [hunk 6]
```diff
@@ -39,24 +42,54 @@ def __init__(self, public_key=None, private_key=None, use_ssl=None,
...
+            if key in request.session:
+                expires_time = request.session[key]
+                return expires_time > time.time()
+
+        return False
+
+    def set_cached_result(self):
+        request = self.get_request_object()
+        if request:
+            key = 'rcf-{path}cached-result'.format(path=request.get_full_path())
```
Comment (adamchainz) [inline captcha/fields.py]: `get_full_path` with key prefix, version, `rcf` etc could easily overrun the 255 char limit on cache keys on memcache

better to put a hash of the URL

**Comment by AltusBarry** [comment]
Superseded by https://github.com/praekelt/django-recaptcha/pull/196

**Pull request closed without merging**